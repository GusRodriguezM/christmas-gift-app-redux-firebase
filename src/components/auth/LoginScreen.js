import React, { useMemo, useEffect } from 'react';
import { Link } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import validator from 'validator';

import { startGoogleSignIn, startLoginWithEmailPassword } from '../../store/slices/auth';
import { removeErrorMessage, setErrorMessage } from '../../store/slices/ui';
import { useForm } from '../../hooks';

import { Button } from '../styles/shared/Button.styled';
import Container from '../styles/auth/Container.styled';
import Input from '../styles/elements/Input.styled';
import Toast from '../styles/Toast/Toast';

export const LoginScreen = () => {

    const { status } = useSelector( state => state.auth );
    const { msgError } = useSelector( state => state.ui );
    const dispatch = useDispatch();

    const [formValues, handleInputChange] = useForm({
        email: '',
        password: ''
    });

    const { email, password } = formValues;

    const isAuthenticating = useMemo(() => status === 'checking', [status]);

    useEffect(() => {
        if(msgError.length > 0){
            Toast.fire({
                icon: 'error',
                title: `${msgError}`
            });
        }
    }, [msgError]);

    const isFormValid = () => {
        if(validator.isEmpty(email) && validator.isEmpty(password)){
            dispatch( setErrorMessage('Email and password are required') );
            return false;
        }else if(validator.isEmpty(email)){
            dispatch( setErrorMessage('The email is required') );
            return false;
        }else if(!validator.isEmail(email)){
            dispatch( setErrorMessage('The email is not valid') );
            return false;
        }else if(validator.isEmpty(password)){
            dispatch( setErrorMessage('The password is required') );
            return false;
        }

        dispatch( removeErrorMessage() );
        return true;
    }

    const handleLogin = (e) => {
        e.preventDefault();
        
        if(isFormValid())
            dispatch( startLoginWithEmailPassword({email, password}) );
    }

    const handleGoogleSignIn = () => {
        dispatch( startGoogleSignIn() );
    }

    return (
        <Container>

            <h1>Login</h1>

            <Container.AuthForm onSubmit={handleLogin}>

                <Input
                    id='name'
                    type='text'
                    placeholder='Email'
                    name='email'
                    autoComplete='off'
                    value={email}
                    onChange={handleInputChange}
                />

                <Input 
                    type='password'
                    placeholder='Password'
                    name='password'
                    autoComplete='off'
                    value={password}
                    onChange={handleInputChange}
                />

                <Button 
                    type='submit'
                    disabled={isAuthenticating}
                    inactive={isAuthenticating}
                >
                    <span>Login</span>
                    <i className="fa-solid fa-user-lock"></i>
                </Button>

                <Button 
                    onClick={handleGoogleSignIn}
                    disabled={isAuthenticating}
                    inactive={isAuthenticating}
                >
                    <span>Google Login</span>
                    <i className="fa-brands fa-google"></i>
                </Button>

            </Container.AuthForm>

            <Link
                aria-disabled={isAuthenticating}
                to='/auth/register'
            >
                Create a new account
            </Link>
        </Container>
    )
}