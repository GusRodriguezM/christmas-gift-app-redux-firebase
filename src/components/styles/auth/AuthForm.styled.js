import styled from 'styled-components';

const AuthForm = styled.form`
    width: 100%;
    height: 60%;
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: space-evenly;
`;

export default AuthForm;