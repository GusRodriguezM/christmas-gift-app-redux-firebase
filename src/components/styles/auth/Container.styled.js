import styled from 'styled-components';

import AuthForm from './AuthForm.styled';

const Container = styled.div`
    width: 450px;
    height: 450px;
    background: rgba( 211, 226, 236, 0.5 );
    box-shadow: 0 8px 32px 0 rgba( 31, 38, 135, 0.37 );
    backdrop-filter: blur(4px);
    border-radius: 10px;
    display: flex;
    flex-direction: column;
    align-items: center;
`;

Container.AuthForm = AuthForm;

export default Container;