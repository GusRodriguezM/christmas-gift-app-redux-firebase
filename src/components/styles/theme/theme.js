const theme = {
    colors: {
        black: '#000000',
        loginButton: '#4285f4',
        disabledButton: '#999595',
        red: '#C30F16',
        white: '#FFFFFF',
        list: '#c8e4e4',
        navbar: '#e8ceac',
        boxShadow: 'rgba(9, 9, 9, 0.23)',
    }
}

export default theme;