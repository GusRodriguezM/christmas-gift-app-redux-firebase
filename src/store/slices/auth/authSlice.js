import { createSlice } from '@reduxjs/toolkit';

export const authSlice = createSlice({
  name: 'auth',
  initialState: {
    status: 'checking', //options: checking, non-authenticated, authenticated
    uid: null,
    email: null,
    userName: null,
    photoURL: null,
    errorMessage: null,
    logged: false
  },
  reducers: {
    login: (state, { payload }) => {
        state.status = 'authenticated';
        state.uid = payload.uid;
        state.email = payload.email;
        state.userName = payload.displayName;
        state.photoURL = payload.photoURL;
        state.errorMessage = payload.errorMessage;
    },
    logout: (state, { payload }) => {
        state.status = 'non-authenticated';
        state.uid = null;
        state.email = null;
        state.userName = null;
        state.photoURL = null;
        state.errorMessage = payload.errorMessage;
    },
    checkingUserCredentials: (state, action) => {
      state.status = 'checking';
    },
  },
});

export const { login, logout, checkingUserCredentials } = authSlice.actions;