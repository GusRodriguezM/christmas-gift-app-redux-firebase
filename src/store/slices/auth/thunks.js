import { signInWithGoogle } from "../../../firebase/providers";
import { checkingUserCredentials, login, logout } from "./"

export const checkingAuthentication = (email, password) => {
    return async( dispatch ) => {
        dispatch( checkingUserCredentials() );
    }
}

export const startGoogleSignIn = () => {
    return async( dispatch ) => {
        dispatch( checkingUserCredentials() );
        const result = await signInWithGoogle();
        
        if(!result.ok)
            return dispatch( logout( result.errorMessage ) );

        dispatch( login(result) );
    }
}