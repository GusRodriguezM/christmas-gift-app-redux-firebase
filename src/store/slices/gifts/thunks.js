import { collection, deleteDoc, doc, setDoc, writeBatch } from 'firebase/firestore/lite';
import { FirebaseDB } from '../../../firebase/config';
import { loadGifts } from '../../../helpers';
import { addGift, setSavingGift, setGifts, editGift, deleteGiftById, cleanList } from './';

export const startAddingNewGift = (newGift) => {
    return async(dispatch, getState) => {

        dispatch( setSavingGift() );
        
        const { uid } = getState().auth;

        const newDoc = doc( collection( FirebaseDB, `${uid}/list/gifts` ) );
        await setDoc( newDoc, newGift );

        newGift.id = newDoc.id;

        dispatch( addGift(newGift) );

    }
}

export const startLoadingGifts = () => {
    return async(dispatch, getState) => {

        const { uid } = getState().auth;
        if(!uid) throw new Error('The user id does not exist');

        const gifts = await loadGifts(uid);

        dispatch( setGifts(gifts) );
    }
}

export const startSavingGift = (giftToEdit) => {
    return async(dispatch, getState) => {
        dispatch( setSavingGift() );

        const { uid } = getState().auth;
        const { activeGift } = getState().gifts;

        const docRef = doc(FirebaseDB, `${uid}/list/gifts/${activeGift.id}`);
        await setDoc(docRef, giftToEdit, { merge: true });

        giftToEdit.id = activeGift.id;
        dispatch( editGift(giftToEdit) );
    }
}

export const startDeletingGift = (id) => {
    return async(dispatch, getState) => {

        const { uid } = getState().auth;

        const docRef = doc(FirebaseDB, `${uid}/list/gifts/${id}`);
        await deleteDoc(docRef);

        dispatch( deleteGiftById(id) );
        
    }
}

export const startDeletingGifts = () => {
    return async(dispatch, getState) => {
        
        //Getting the user id and the gifts array from the store
        const { uid } = getState().auth;
        const { gifts } = getState().gifts;

        //We use batch writes. Basically this instruction allows to execute a set of different operations
        const batch = writeBatch(FirebaseDB);

        //Loop over the array of gifts to delete. We send the reference to the doc
        gifts.forEach(gift => {
            batch.delete( doc(FirebaseDB, `${uid}/list/gifts/${gift.id}`) );
        });

        //Finally, we commit the batch
        await batch.commit();

        //We clean the array in the store
        dispatch( cleanList() );

    }
}