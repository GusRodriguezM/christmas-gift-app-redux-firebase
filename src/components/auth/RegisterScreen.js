import React from 'react';
import { useNavigate } from 'react-router-dom';
import { useDispatch } from 'react-redux';

import { login } from '../../store/slices/auth/authSlice';
import { useForm } from '../hooks/useForm';
import Container from '../styles/auth/Container.styled';
import Input from '../styles/elements/Input.styled';
import { Button } from '../styles/shared/Button.styled';

export const RegisterScreen = () => {

    const dispatch = useDispatch();
    const navigate = useNavigate();

    const [formValues, handleInputChange] = useForm({
        name: '',
        email: '',
        password: ''
    });

    const { name, email, password } = formValues;

    const handleLogin = (e) => {
        e.preventDefault();
        
        dispatch( login(email) );
        navigate('/', {
            replace: true
        });
    }

    const handleNavigate = () => {
        navigate('/auth/login', {
            replace: true
        });
    }

    return (
        <Container>

            <h1 className='login__title'>Sign in</h1>

            <Container.AuthForm onSubmit={handleLogin} className='login__form'>

                <Input
                    type='text'
                    placeholder='Name'
                    name='name'
                    className='login__form--input'
                    autoComplete='off'
                    value={name}
                    onChange={handleInputChange}
                />

                <Input 
                    type='text'
                    placeholder='Email'
                    name='email'
                    className='login__form--input'
                    autoComplete='off'
                    value={email}
                    onChange={handleInputChange}
                />

                <Input 
                    type='password'
                    placeholder='Password'
                    name='password'
                    className='login__form--input'
                    autoComplete='off'
                    value={password}
                    onChange={handleInputChange}
                />

                <Button type='submit'>
                    <span>Login</span>
                </Button>

            </Container.AuthForm>

            <Container.Span>I already have an account</Container.Span>
            <Button
                onClick={handleNavigate}
            >
                Login
            </Button>
        </Container>
    )
}
