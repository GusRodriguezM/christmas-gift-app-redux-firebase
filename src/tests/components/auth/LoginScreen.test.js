import { Provider } from "react-redux";
import { configureStore } from "@reduxjs/toolkit";
import { ThemeProvider } from "styled-components";
import { MemoryRouter } from "react-router-dom";
import { fireEvent, render, screen } from "@testing-library/react";
import { LoginScreen } from "../../../components/auth/LoginScreen";
import { authSlice } from "../../../store/slices/auth";
import { startGoogleSignIn, startLoginWithEmailPassword } from "../../../store/slices/auth/thunks";
import { uiSlice } from "../../../store/slices/ui";
import { nonAuthenticatedState } from "../../fixtures/authFixtures";
import theme from "../../../components/styles/theme/theme";
import 'jest-styled-components';

//Making mocks of the functions
const mockStartGoogleSignIn = jest.fn();
const mockStartLoginWithEmailPassword = jest.fn();

//Declaring the route of the mocks and for the second passing props like in the login page
jest.mock('../../../store/slices/auth/thunks', () => ({
    startGoogleSignIn: () => mockStartGoogleSignIn,
    startLoginWithEmailPassword: ({email, password}) => {
        return () => mockStartLoginWithEmailPassword({email, password})
    },
}));

//Mocking the store. Using the auth and ui reducer and setting a specific state to the auth
const store = configureStore({
    reducer: {
        auth: authSlice.reducer,
        ui: uiSlice.reducer,
    },
    preloadedState: {
        auth: nonAuthenticatedState
    },
});

describe('Tests in the <LoginScreen />', () => {

    //Cleaning the mocks
    beforeEach( () => jest.clearAllMocks() );

    test('Should render the component correctly and match with the snapshot', () => {

        //Rendering the component with the router, a provider for redux and for the styles to the styled components
        const { container } = render(
            <Provider store={ store } >
                <MemoryRouter>
                    <ThemeProvider theme={ theme } >
                        <LoginScreen />
                    </ThemeProvider>
                </MemoryRouter>
            </Provider>
        );

        //Finally we expect that the render matches with the snapshot
        expect( container ).toMatchSnapshot();
        
    });
    
    test('<LoginScreen /> should have all the elements of the form', () => {

        //Rendering the component
        render(
            <Provider store={ store } >
                <MemoryRouter>
                    <ThemeProvider theme={ theme } >
                        <LoginScreen />
                    </ThemeProvider>
                </MemoryRouter>
            </Provider>
        );
        
        //Expecting that the component have the elements for an input, link, heading and button
        expect( screen.getAllByText('Login').length ).toBeGreaterThanOrEqual(1);
        expect( screen.getAllByRole('textbox').length ).toBe(1);
        expect( screen.getAllByRole('link').length ).toBe(1);
        expect( screen.getAllByRole('button').length ).toBe(2);

    });

    test('The Google button should call startGoogleSignIn', () => {

        //Rendering the component
        render(
            <Provider store={ store } >
                <MemoryRouter>
                    <ThemeProvider theme={ theme } >
                        <LoginScreen />
                    </ThemeProvider>
                </MemoryRouter>
            </Provider>
        );

        //Getting the reference to the button by a label
        const googleBtn = screen.getByLabelText('googleBtn');
        
        //Firing the click event
        fireEvent.click( googleBtn );

        //Finally we expect that the function to start the login with Google have been called
        expect( mockStartGoogleSignIn ).toHaveBeenCalled();

    });

    test('The login button should submit the form and call startLoginWithEmailPassword with specific values', () => {

        //Constants to send to the form
        const email = 'bruce.wayne@gmail.com';
        const password = '123456';

        //Rendering the component
        render(
            <Provider store={ store } >
                <MemoryRouter>
                    <ThemeProvider theme={ theme } >
                        <LoginScreen />
                    </ThemeProvider>
                </MemoryRouter>
            </Provider>
        );
        
        //Getting the reference to the input (for the email)
        const emailField = screen.getByPlaceholderText('Email');
        
        //Firing an event to put the value of the variable into the input
        fireEvent.change( emailField, { target: { name: 'email', value: email } } );

        //Getting the reference to the input (for the password)
        const passwordField = screen.getByPlaceholderText('Password');

        //Firing an event to put the value of the variable into the input
        fireEvent.change( passwordField, { target: { name: 'password', value: password } } );

        //Getting the reference to the form
        const loginForm = screen.getByLabelText('login-form');

        //Firing the submit
        fireEvent.submit(loginForm);

        //Expecting that the login with email and password have been called with the expected values
        expect( mockStartLoginWithEmailPassword ).toHaveBeenCalledWith( {email, password} );

    });

    test('Sweetalert should show if a field in the form is incorrect', () => {

        //Variables for the form
        const email = 'bruce.waynegmail.com';
        const password = '123456';

        //Rendering the component
        render(
            <Provider store={ store } >
                <MemoryRouter>
                    <ThemeProvider theme={ theme } >
                        <LoginScreen />
                    </ThemeProvider>
                </MemoryRouter>
            </Provider>
        );
        
        //Getting the reference for the input (email)
        const emailField = screen.getByPlaceholderText('Email');
        
        //Firing the event to change the value of the input
        fireEvent.change( emailField, { target: { name: 'email', value: email } } );

        //Getting the reference for the input (password)
        const passwordField = screen.getByPlaceholderText('Password');

        //Firing the event to change the value of the input
        fireEvent.change( passwordField, { target: { name: 'password', value: password } } );

        //Getting the reference to the form
        const loginForm = screen.getByLabelText('login-form');

        //Firing the submit of the form
        fireEvent.submit(loginForm);

        //Expecting that the sweetalert notification appears because the email is wrong
        expect( screen.getAllByRole('heading', { name: 'The email is not valid' }).length ).toBe(1);

    });

});