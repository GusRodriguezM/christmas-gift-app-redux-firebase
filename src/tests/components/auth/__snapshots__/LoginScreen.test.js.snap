// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in the <LoginScreen /> Should render the component correctly and match with the snapshot 1`] = `
.c3 {
  padding: 5px 5px;
  margin: 5px 0;
  text-align: center;
  color: #FFFFFF;
  border-radius: 5px;
  cursor: pointer;
  border: none;
  line-height: 1;
  width: 100px;
  background-color: #4285f4;
}

.c1 {
  width: 100%;
  height: 80%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: space-evenly;
  -webkit-justify-content: space-evenly;
  -ms-flex-pack: space-evenly;
  justify-content: space-evenly;
}

.c0 {
  width: 450px;
  height: 450px;
  background: rgba( 211,226,236,0.5 );
  box-shadow: 0 8px 32px 0 rgba( 31,38,135,0.37 );
  -webkit-backdrop-filter: blur(4px);
  backdrop-filter: blur(4px);
  border-radius: 10px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-top: 60px;
}

.c2 {
  margin: 10px 0;
  width: 95%;
}

@media screen and (max-width:425px) {
  .c3 span {
    display: none;
  }
}

@media screen and (min-width:426px) and (max-width:1440px) {
  .c3 i {
    display: none;
  }
}

@media screen and (max-width:375px) {
  .c3 {
    width: 50px;
  }
}

@media screen and (min-width:376px) and (max-width:1024px) {
  .c3 {
    width: 80px;
  }
}

@media screen and (max-width:375px) {
  .c0 {
    width: 400px;
  }
}

<div>
  <div
    class="c0"
  >
    <h1>
      Login
    </h1>
    <form
      aria-label="login-form"
      class="c1"
    >
      <input
        autocomplete="off"
        class="c2"
        name="email"
        placeholder="Email"
        type="text"
        value=""
      />
      <input
        autocomplete="off"
        class="c2"
        name="password"
        placeholder="Password"
        type="password"
        value=""
      />
      <button
        class="c3"
        type="submit"
      >
        <span>
          Login
        </span>
        <i
          class="fa-solid fa-user-lock"
        />
      </button>
      <button
        aria-label="googleBtn"
        class="c3"
      >
        <span>
          Sign in with Google
        </span>
        <i
          class="fa-brands fa-google"
        />
      </button>
    </form>
    <a
      aria-disabled="false"
      href="/auth/register"
    >
      Create a new account
    </a>
  </div>
</div>
`;
